### DISCUSSION TITLE
Does the JWT role claim have anything to do with postgresql roles?

### DISCUSSION THREAD

**Discussion by mstade** [supabase/supabase#596] [Questions] [answered]
Hi!

I'm still trying to figure out how we should implement our custom auth (see #544) and I've been trying to reverse engineer how the JWT tokens are generated. It seems to me that so long as we include the `aud`, `exp` and `sub` claims we should be alright. There are however some non-standard claims in the tokens issued by Supabase, such as `email`, `app_metadata`, `user_metadata` and `role`. The last one is particularly interesting, as it seems to always be set to `authenticated` (as is `aud`.)

Now, I've been looking through the `auth` schema and see there's a `role` function in there, which seems to be a convenience function for picking out data from the JWT. However, I can't seem to find any details on this function. Is it at all related to postgresql roles, or is this a case of name confusion? There is of course an `authenticated` role in the database, which makes me think that maybe it does matter, and there's also the `aud` and `role` columns in the `auth` table, but I can't seem to find any documentation on whether or how these should be used.

What we want to do in our app is assign different roles to people depending on their access, and it certainly wouldn't hurt if we could use convenience functions like `auth.role()` in our policies. We could of course just create our own convenience functions for this, but I'm wondering if there's any plan behind `auth.role()` and if you mean for it to be used in RLS policies the same way you suggest using `auth.uid()` for instance.

Anyway, I guess the question is – does the `role` claim in the JWT have anything to do with postgresql roles or should we just always set this to `authenticated` when generating our JWTs?

**Comment by mstade** [comment, marked as answer]
Turns out: yes it does! In fact, as far as I can tell between the `role` and the `sub` claims it seems the authentication and authorization details are met. I’m not sure whether the `aud` claim matters to supabase, it seemed to not make much of a difference but I didn’t test it very thoroughly. Obviously you’ll need `exp` etc or else supabase (postgrest?) will complain. [JWT.io](https://jwt.io/) has been a great resource for testing since you can just mess with the claim values and use the JWT secret from the dashboard to properly sign the token.

So it seems the role claim is indeed used for mapping to Postgres roles, which presumably means if we create new roles in the database we could use those with the tokens. I’m thinking this could be useful for sudo like functionality where in our app we’d ask a token service for a privileged token with the necessary role set to perform a given action, and a much lower `exp` claim (i.e. token expiration) of say 60 seconds rather than the default 3600, or whatever makes sense.

I did a test with a database function that takes a subject as the input, and checks `auth.uid()` against a list of roles to see if the current user had the role (basically look up roles in a list mapped to the user id) to request a new token for a given user, like a service being allowed to impersonate a user. It worked, but only if I use the secret service key, and set the role to `supabase_admin`. If I tried using the anon key I got an error like this:


```postgresql
ERROR:  permission denied to set role "supabase_admin"
```

Which sounds like it makes sense, you probably don’t want the anon (i.e. public?) key to be used for elevating privileges, but my postgresql-fu is really basic and I couldn’t figure out how to make this work for *any* role, let alone `supabase_admin`. I figure if I I can sort that out I can make a role that’s limited to read only reads of the `auth` table, such that I can match an input `subject` uuid against an actual user, look up whether they have the requested role in its list of assigned roles,  and then return an appropriate JWT. Obviously the function would also check `auh.uid()` so that we can verify that the current (service) account has permission to request a token.


I’ve [googled this](https://stackoverflow.com/questions/33434907/postgresql-how-to-grant-permission-for-set-role) but I can’t seem to figure out how to get past the aforementioned error. Anyone have any thoughts?

**Reply by steve-chavez** [reply]
Hey @mstade,

Yes, the `jwt` role claim is mapped to a database role.

> ERROR:  permission denied to set role "supabase_admin"

supabase_admin is a highly privileged role(SUPERUSER), so I wouldn't recommend using it for the API.

However if you create another role such as `web_admin`, you can clear that error by doing:

```sql
GRANT web_admin to authenticator;
```

For more details, check https://postgrest.org/en/v7.0.0/auth.html.

**Reply by mstade** [reply]
Absolutely, I never indented to actually use the `supabase_admin` role for anything – it was just to test some hypotheses. Thank you so much for the link, I've only had a cursory look at it but it seems to be exactly the kind of information I'm looking for. Right there in the introduction it says:

> The authenticator should be created NOINHERIT and configured in the database to have very limited access. It is a chameleon whose job is to “become” other users to service authenticated HTTP requests. The picture below shows how the server handles authentication. If auth succeeds, it switches into the user role specified by the request, otherwise it switches into the anonymous role.

This is of course exactly what I'm trying to achieve, so hopefully these docs should put me on the right track – thank you so much!

Also, please forgive my ignorance. It's a bit tricky when you don't know the components to then also know where to go look for more information. (Postgrest for example is completely new to me.) It's very much a learning experience, and I'm _very_ grateful to you for lending your time to answer these probably pretty ignorant questions. Thank you so much @steve-chavez, you've been extremely helpful! 🙏

**Reply by steve-chavez** [reply]
@mstade No problem! Glad to help!

**Reply by eifo** [reply]
> Hey @mstade,
> 
> Yes, the `jwt` role claim is mapped to a database role.
> 
> > ERROR:  permission denied to set role "supabase_admin"
> 
> supabase_admin is a highly privileged role(SUPERUSER), so I wouldn't recommend using it for the API.
> 
> However if you create another role such as `web_admin`, you can clear that error by doing:
> 
> ```sql
> GRANT web_admin to authenticator;
> ```
> 
> For more details, check https://postgrest.org/en/v7.0.0/auth.html.

@steve-chavez after granting "new role" to authenticator how would we go to actually assign this "new role" role to a specific user in auth.users instead of the current "authenticated" role, I tried update auth.users set role = 'new role' but it does not work with RLS when trying to access auth.role() = 'new role'

**Reply by steve-chavez** [reply]
@eifo That's configured by the [JWT_AUD](https://github.com/supabase/gotrue/#json-web-tokens-jwt) in [gotrue](https://github.com/supabase/gotrue/), right now I think it's not configurable at runtime. Perhaps can you open an issue there and ask about adding the enhancement?
